Decode old kyber swaps

yabirgb opened the pull request.

LefterisJP commented on rotkehlchen/chain/ethereum/decoding/kyber/decoder.py:
```diff
@@ -0,0 +1,73 @@
+    from rotkehlchen.chain.ethereum.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.ethereum.manager import EthereumManager
+
+TOKEN_SWAP = b"\x18I\xbdj\x03\n\x1b\xca(\xb847\xfd=\xe9o='\xa5\xd1r\xfa~\x9cx\xe7\xb6\x14h\x92\x8a9"  # noqa: E501
```
Which event is this? I don't see an even log with the name TokenSwap here: https://etherscan.io/tx/0x615c28b9867c786d7365a7d60f90a68b95cc2acd025fb0f96e22134e8dc2602f#eventlog

LefterisJP commented on rotkehlchen/chain/ethereum/decoding/kyber/decoder.py:
```diff
@@ -0,0 +1,73 @@
+            all_logs: List[EthereumTxReceiptLog],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        source_token_address = hex_or_bytes_to_address(tx_log.data[:32])
+        source_token = ethaddress_to_asset(source_token_address)
```
no this is not enough. You should find the exact same event. Which means token, amount, source and generally verify as many ofthe fiellds as possible.

Also where is the destination/to token?

Check out the uniswap or 1inch-v1 and 1inch-v2 swap decoding to understand.

Your example here would fail here and pick any random ERC transfer for the "receive" part and also for the spend part if amount, or address is not the same

LefterisJP commented on rotkehlchen/chain/ethereum/decoding/kyber/decoder.py:
```diff
@@ -0,0 +1,86 @@
+            action_items: Optional[List[ActionItem]],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        if tx_log.topics[0] == EXECUTE_TRADE:
+            return self._decode_old_swap(tx_log=tx_log, decoded_events=decoded_events)
```
nitpick: should have the name of the event you are decoding

LefterisJP commented on rotkehlchen/chain/ethereum/decoding/kyber/decoder.py:
```diff
@@ -0,0 +1,86 @@
+        spent_amount_raw = hex_or_bytes_to_int(tx_log.data[64:96])
+        return_amount_raw = hex_or_bytes_to_int(tx_log.data[96:128])
+        spent_amount = asset_normalized_value(amount=spent_amount_raw, asset=source_token)
+        return_amount = asset_normalized_value(amount=return_amount_raw, asset=destination_token)
```
Why not use the `KyberTrade` event instead? That one has more information which is also needed. Specifically when comparing them it seems to be the only one that has a destination address for the result of the swap.

![2022-02-22_09-25](https://user-images.githubusercontent.com/1658405/155092335-10d5a650-d88f-4997-9b99-db40b16d0c64.png)


 If you also check the contract and read its code you will see that the destination address of the swap is also configurable.

![2022-02-22_09-28](https://user-images.githubusercontent.com/1658405/155092585-4cd2f3d9-f639-4a70-805c-4c7f27418d2f.png)

 So your current solution, using that event is going to miss all cases where the destination of the swap is not the same as the sender.

yabirgb replied:
I didn't use that one because when I inspected the topics in the decode action I got only 2
````
[b"\x18I\xbdj\x03\n\x1b\xca(\xb847\xfd=\xe9o='\xa5\xd1r\xfa~\x9cx\xe7\xb6\x14h\x92\x8a9", b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00rw\xf7\x84\x99fBm4]\x8fk\x9a\xfd\x1d=\x89\x180\x83']
````

The first one is the one I used `execute trade` and the other I didn't find in the list

```
>>> bytearray(b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00rw\xf7\x84\x99fBm4]\x8fk\x9a\xfd\x1d=\x89\x180\x83").hex()
'0000000000000000000000007277f7849966426d345d8f6b9afd1d3d89183083'
```

LefterisJP replied:
It's there. If you don't find it you are doing something wrong. The transaction has 8 logs

LefterisJP commented on rotkehlchen/chain/ethereum/decoding/kyber/decoder.py:
```diff
@@ -0,0 +1,86 @@
+            if event.event_type == HistoryEventType.SPEND and event.location_label == sender and event.asset == source_token and event.balance.amount == spent_amount:  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.counterparty = 'kyber'
```
I don't know how kyber is versioning itself but it woudl be nice to use some kind of version. Just like we do in uniswap, 1inch etc.

yabirgb commented:
I've changed the event used and added both legacy contracts:

- As reference for legacy contract upgraded version I took https://etherscan.io/tx/0xe80928d5e21f9628c047af1f8b191cbffbb6b8b9945adb502cfb3af152552f22
- As reference for legacy contract old version I took https://etherscan.io/tx/0x615c28b9867c786d7365a7d60f90a68b95cc2acd025fb0f96e22134e8dc2602f

LefterisJP commented on rotkehlchen/chain/ethereum/decoding/kyber/decoder.py:
```diff
@@ -0,0 +1,95 @@
+
+    def addresses_to_decoders(self) -> Dict[ChecksumEthAddress, Tuple[Any, ...]]:
+        return {
+            string_to_ethereum_address('0x9ae49C0d7F8F9EF4B864e004FE86Ac8294E20950'): (self.decode_action,),  # noqa: E501
```
Hmmm this is not the correct approach.

The address in question is what should map to the decoder only for that address.

From what I understand here you have 2 different contracts.

1. First name them. Add them as named constants in the file
2. Each address should map to a different decoding functon.

LefterisJP reviewed (commented):
Can you also start adding the unit tests as discused? It can happen in another PR as long as you fix the problems with this one

LefterisJP commented on rotkehlchen/chain/ethereum/decoding/kyber/decoder.py:
```diff
@@ -0,0 +1,125 @@
+
+        return None, None
+
+    def _decode_legacy_upgraded_trade(  # pylint: disable=no-self-use
```
Can't you create a common method called by oth of these? A lot of code from these 2 functions can be abstracted to a common one, right?

LefterisJP commented on rotkehlchen/tests/unit/decoders/kyber.py:
```diff
@@ -0,0 +1,223 @@
+
+@pytest.mark.parametrize('ethereum_accounts', [['0x6d379cb5BA04c09293b21Bf314E7aba3FfEAaF5b']])  # noqa: E501
+def test_kyber_legacy_old_contract(database, ethereum_manager):
+    # Data for trade
+    # https://etherscan.io/tx/0xe9cc9f27ef2a09fe23abc886a0a0f7ae19d9e2eb73663e1e41e07a3e0c011b87
```
Use a docstring here
```suggestion
    """Data for trade taken from:
    https://etherscan.io/tx/0xe9cc9f27ef2a09fe23abc886a0a0f7ae19d9e2eb73663e1e41e07a3e0c011b87
    """
```

LefterisJP commented on rotkehlchen/tests/unit/decoders/kyber.py:
```diff
@@ -0,0 +1,223 @@
+    # https://etherscan.io/tx/0xe9cc9f27ef2a09fe23abc886a0a0f7ae19d9e2eb73663e1e41e07a3e0c011b87
+    cursor = database.conn.cursor()
+    msg_aggregator = MessagesAggregator()
+    tx_hex = '0xe9cc9f27ef2a09fe23abc886a0a0f7ae19d9e2eb73663e1e41e07a3e0c011b87'
```
for this use deserialize_evm_tx_hash() from types

yabirgb replied:
It was used below but don't know why I was using hexbytes for the transaction hash. Using it everywhere where is needed

LefterisJP commented on rotkehlchen/tests/unit/decoders/kyber.py:
```diff
@@ -0,0 +1,223 @@
+            identifier=3,
+            extras=None,
+        )]
+    assert events == expected_events
```
Hmm the order seems wrong. Shouldn't they be ordered by sequence index?

LefterisJP commented on rotkehlchen/tests/unit/decoders/kyber.py:
```diff
@@ -0,0 +1,223 @@
+
+@pytest.mark.parametrize('ethereum_accounts', [['0x5340F6faff9BF55F66C16Db6Bf9E020d987F87D0']])  # noqa: E501
+def test_kyber_legacy_new_contract(database, ethereum_manager):
+    # Data for trade
+    # https://etherscan.io/tx/0xe80928d5e21f9628c047af1f8b191cbffbb6b8b9945adb502cfb3af152552f22
```
use docstring

LefterisJP commented on rotkehlchen/tests/unit/decoders/kyber.py:
```diff
@@ -0,0 +1,223 @@
+    # https://etherscan.io/tx/0xe80928d5e21f9628c047af1f8b191cbffbb6b8b9945adb502cfb3af152552f22
+    cursor = database.conn.cursor()
+    msg_aggregator = MessagesAggregator()
+    tx_hex = '0xe80928d5e21f9628c047af1f8b191cbffbb6b8b9945adb502cfb3af152552f22'
```
same comment as other function

LefterisJP commented on rotkehlchen/tests/unit/decoders/kyber.py:
```diff
@@ -0,0 +1,223 @@
+    cursor.executemany(
+        'INSERT INTO ethtx_receipt_log_topics (tx_hash, log_index, topic, topic_index) '
+        'VALUES(? ,? ,?, ?)',
+        [(b'\xe8\t(\xd5\xe2\x1f\x96(\xc0G\xaf\x1f\x8b\x19\x1c\xbf\xfb\xb6\xb8\xb9\x94Z\xdbP,\xfb:\xf1RU/"', 349, b'\xdd\xf2R\xad\x1b\xe2\xc8\x9bi\xc2\xb0h\xfc7\x8d\xaa\x95+\xa7\xf1c\xc4\xa1\x16(\xf5ZM\xf5#\xb3\xef', 0), (b'\xe8\t(\xd5\xe2\x1f\x96(\xc0G\xaf\x1f\x8b\x19\x1c\xbf\xfb\xb6\xb8\xb9\x94Z\xdbP,\xfb:\xf1RU/"', 349, b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00S@\xf6\xfa\xff\x9b\xf5_f\ ...
```
Here and before you seem to be putting a hell of a lot of log events for the transaction. Save space and only put the relevant ones that the test is going to decode. This takes too much space as you currently have it.

yabirgb replied:
I believed it was needed to correctly decode the send/receive events as when I removed part of them I was loosing this events. I've removed most of the data and only left the methods that do the swaps and the send of the tokens

LefterisJP reviewed (approved):
LGTM
